PyRoute/TradeCodes.py

```diff
@@ -80,6 +111,123 @@ def __init__(self, initial_codes):
         self.dcode = list(self.dcode) + self.colony + self.owner
         self.dcode = sorted(self.dcode)
 
+    def _preprocess_initial_codes(self, initial_codes):
+        raw_codes = initial_codes.split()
+        # look for successive codes that should be part of the same name (eg  "(Carte" , then "Blanche)", becoming
+        # "(Carte Blanche)", and bolt them together
+        num_codes = len(raw_codes)
+        codes = []
+        for i in range(0, num_codes):
+            raw = raw_codes[i]
+            # Filter duplicates
+            if raw in codes:
+                continue
+            if '' == raw:
+                continue
+            if ')' == raw:
+                continue
+            if raw.startswith('Di('):
+                codes.append(raw)
+                continue
+            if 7 < len(raw) and '(' == raw[0] and ')' == raw[-2]:  # Let older-style sophont codes through
+                codes.append(raw)
+                continue
+            if 2 == len(raw) and ('W' == raw[1] or raw[1].isdigit()):
+                if 'C' == raw[0]:  # Compact chirper population
+                    pop = raw[1]
+                    codes.append('Chir' + pop)
+                    continue
+                if 'D' == raw[0]:  # Compact Droyne population
+                    pop = raw[1]
+                    codes.append('Droy' + pop)
+                    continue
+            if not raw.startswith('(') and '(' in raw and raw.endswith(')'):
+                continue
+            if not raw.endswith(')') and ')' in raw and raw.startswith('(') and 7 > len(raw):
+                continue
+            if not raw.startswith('(') and not raw.endswith(')') and '(' in raw and ')' in raw:
+                continue
+            if 7 == len(raw) and '(' == raw[0] and ')' == raw[5]:  # Let preprocessed sophont codes through
+                codes.append(raw)
+                continue
+            if not raw.startswith('('):
+                codes.append(raw)
+                continue
+            if raw.endswith(')'):
+                codes.append(raw)
+                continue
+            if i < num_codes - 1:  # Try to join cases like '(Carte' and 'Blanche)'
+                next = raw_codes[i + 1]
+                if next.endswith(')'):
+                    combo = raw + ' ' + next
+                    codes.append(combo)
+                    raw_codes[i + 1] = ''
+                continue
+
+        initial_codes = ' '.join(codes)
+
+        return codes, initial_codes
+
+    def _process_homeworld(self, homeworld, homeworlds_found, initial_codes):
+        full_name = re.sub(r'\(([^)]+)\)\d?', r'\1', homeworld)
+        homeworlds_found.append(homeworld)
+        match = re.match(r'\w{,2}\(([^)]{,4})[^)]*\)(\d|W|\?)?', homeworld)
```

Comment:
Since this is going to be used for every single world read in (along with others) it may be useful to compile this. see https://docs.python.org/3/library/re.html#re.compile
```suggestion
        search = re.compile(r'\w{,2}\(([^)]{,4})[^)]*\)(\d|W|\?)?') # Goes in the class definition
        match = search.match(homeworld)
```

Reply:
Compiled and hoisted